overlord/servicestate/quota_control.go

```diff
@@ -0,0 +1,254 @@
+		}
+	} else {
+		// make a new group
+		grp, err = quota.NewGroup(name, memoryLimit)
+		if err != nil {
+			return err
+		}
+	}
+
+	// put the snaps in the group
+	grp.Snaps = snaps
+
+	updatedGrps := []*quota.Group{grp}
+	if parentName != "" {
+		updatedGrps = append(updatedGrps, parentGrp)
+	}
+
+	// update the modified groups in state
+	allGrps, err = PatchQuotasState(st, updatedGrps...)
+	if err != nil {
+		return err
+	}
+
+	// ensure the snap services with the group
+	if err := ensureSnapServicesForGroup(st, grp, allGrps); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RemoveQuota deletes the specific quota group.
+// TODO: currently this only supports removing leaf sub-group groups, it doesn't
+// support removing parent quotas, but probably it makes sense to allow that too
+func (mgr *ServiceManager) RemoveQuota(name string) error {
+	st := mgr.state
+	st.Lock()
+	defer st.Unlock()
+
+	allGrps, err := AllQuotas(st)
+	if err != nil {
+		return err
+	}
+
+	// first get the group for later before it is deleted from state
+	grp, ok := allGrps[name]
+	if !ok {
+		return fmt.Errorf("cannot remove non-existent quota group %q", name)
+	}
+
+	// XXX: remove this limitation eventually
+	if len(grp.SubGroups) != 0 {
+		return fmt.Errorf("cannot remove quota group with sub-groups, remove the sub-groups first")
+	}
+
+	// now delete the group from state - do this first for convenience to ensure
+	// that we can just use SnapServiceOptions below and since it operates via
+	// state, it will immediately reflect the deletion
+	delete(allGrps, name)
+	st.Set("quotas", allGrps)
+
+	if err := ensureSnapServicesForGroup(st, grp, allGrps); err != nil {
+		return err
+	}
+
+	// separately delete the slice unit, EnsureSnapServices does not do this for
+	// us
+	if err := wrappers.RemoveQuotaGroup(grp, progress.Null); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func validateSnapForAddingToGroup(st *state.State, name string, allGrps map[string]*quota.Group) error {
+	// validate that the snap exists
+	_, err := snapstate.CurrentInfo(st, name)
+	if err != nil {
+		return err
+	}
+
+	// check that the snap is not already in a group
+	for _, grp := range allGrps {
+		if strutil.ListContains(grp.Snaps, name) {
+			return fmt.Errorf("snap already in quota group %q", grp.Name)
+		}
+	}
+	return nil
+}
+
+// QuotaGroupUpdate reflects all of the modifications that can be performed on
+// a quota group in one operation.
+type QuotaGroupUpdate struct {
+	// AddSnaps is the set of snaps to add to the quota group. These are
+	// instance names of snaps, and are appended to the existing snaps in
+	// the quota group
+	AddSnaps []string
+
+	// NewMemoryLimit is the new memory limit to be used for the quota group. If
+	// zero, then the quota group's memory limit is not changed.
+	NewMemoryLimit quantity.Size
+}
+
+// UpdateQuota updates the quota as per the options.
+func (mgr *ServiceManager) UpdateQuota(name string, updateOpts QuotaGroupUpdate) error {
```

Comment:
Hmm I guess the only thing I can immediately think of that might be useful for the MVP that's not present here (and actually isn't present in my version either tbh) is removing snaps from a quota without removing the quota. But we can accomplish the same thing by just deleting the quota and then recreating it without the snap that is desired to be removed so probably not necessary for now

Reply:
Yeah, I like the idea but IMHO we should create a followup if possible with that feature

Reply:
agreed